servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,624 @@
+                // demand may be distributed than is requested from downstream in order to avoid deadlock scenarios.
+                // To accommodate for the "approximate" mapped demand we maintain a signal queue (bounded by the
+                // concurrency). This presents an opportunity to decouple downstream requestN requests from iterating
+                // all active mapped sources and instead optimistically give out demand here and replenish demand after
+                // signals are delivered to the downstream subscriber (based upon available demand is available).
+                parent.distributeMappedDemand(this);
+            }
+
+            @Override
+            public void onNext(@Nullable final R r) {
+                parent.tryEmitItem(wrapNull(r), this);
+                final int pendingDemand = pendingDemandUpdater.decrementAndGet(this);
+                if (pendingDemand == 0) {
+                    // Emit this item to signify this Subscriber is hungry for more demand when it is available.
+                    parent.tryEmitItem(this, this);
+                } else if (pendingDemand < 0) {
+                    throw new IllegalStateException("Too many onNext signals for Subscriber: " + this +
+                            " pendingDemand: " + pendingDemand);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                if (!parent.source.delayError) {
+                    // Make sure errors aren't delivered out of order relative to onNext signals which maybe queued.
+                    try {
+                        parent.doCancel(true, false);
+                    } finally {
+                        parent.tryEmitItem(TerminalNotification.error(t), this);
+                    }
+                } else {
+                    CompositeException de = parent.delayedError;
+                    if (de == null) {
+                        de = new CompositeException(t);
+                        if (!delayedErrorUpdater.compareAndSet(parent, null, de)) {
+                            de = parent.delayedError;
+                            assert de != null;
+                            de.add(t);
+                        }
+                    } else {
+                        de.add(t);
+                    }
+                    if (parent.removeSubscriber(this, pendingDemandUpdater.getAndSet(this, -1))) {
+                        parent.enqueueAndDrain(TerminalNotification.error(de));
+                    } else {
+                        parent.tryEmitItem(MAPPED_SOURCE_COMPLETE, this);
+                    }
+                }
+            }
+
+            @Override
+            public void onComplete() {
+                if (parent.removeSubscriber(this, pendingDemandUpdater.getAndSet(this, -1))) {
+                    parent.enqueueAndDrain(complete());
+                } else {
+                    parent.tryEmitItem(MAPPED_SOURCE_COMPLETE, this);
```

Comment:
`MAPPED_SOURCE_COMPLETE` going through the queue isn't for concurrency control on the `Subscription` but instead to bound how much data we queue. In order to avoid deadlock scenarios we always give some demand to mapped sources which means queued data should be expected in low downstream demand scenarios. Due to this we simplify demand management and request more from the mapped source in its subscriber's `onSubscribe` method (as opposed to trying to avoid distributing demand when downstream demand is `0`). If we don't go through the queue here we will have to adjust when we requestN flow. I'll add a unit test to demonstrate the behavior and we can consider changes to control flow in a followup PR.

Reply:
discussed offline, fine as is.